### PR TITLE
overlay: use the __symbols__ node in DTB to help overlay generation

### PR DESCRIPTION
**Pull request by ldts**
Summary
-------------
The following patchset reads the `__symbols__` node in a DTB passed in bootargs to help generating the kernel DTB overlay and avoid having to hardcode node paths (node's path-name are not stable and may change thus requiring firmware updates to handle those changes)

**Tested in imx8m - preparing changes for upstreaming to the imx-atf project**

iMX-ATF changes required to support this feature on iMX8m
--------------------------------------------------------------------------------
Trusted Firmware for iMX: https://source.codeaurora.org/external/imx/imx-atf

The imx8mm boot sequence is as follows:
imx8mm

The current imx8m  TF-A 1) does not expect any params from SPL 2) therefore [hardcodes](https://source.codeaurora.org/external/imx/imx-atf/commit/?h=imx_5.4.3_2.0.0&id=f1a195b5cce64365a7227557a9009a4f545aa02d) the addresses where the FDT overlay will be located for OP-TEE (BL32) and U-Boot (BL33) 3) and clears the FDT overlay memory region

This could be modified so that SPL handles the DTB it reads from a FIT to TF-A:

```
diff --git a/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c b/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
index 049192880..66af6de5f 100644
--- a/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
+++ b/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
@@ -176,7 +176,10 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
        /* Make sure memory is clean */
        mmio_write_32(BL32_FDT_OVERLAY_ADDR, 0);
        bl33_image_ep_info.args.arg3 = BL32_FDT_OVERLAY_ADDR;
-   bl32_image_ep_info.args.arg3 = BL32_FDT_OVERLAY_ADDR;
+ /* bl32 could be receiving a DTB on input while also asked to generate
+  * an overlay
+  */
+ bl32_image_ep_info.args.arg3 = arg1 > 0 ? arg1 : BL32_FDT_OVERLAY_ADDR;
 #endif
 #endif

```

Then to actually receive the arguments TF-A could do:

1) if the HW/Bootrom allows for it, do the reset actions in SPL and then

```
diff --git a/plat/imx/imx8m/imx8mm/platform.mk b/plat/imx/imx8m/imx8mm/platform.mk
index 600a3c449..0161ff8e2 100644
--- a/plat/imx/imx8m/imx8mm/platform.mk
+++ b/plat/imx/imx8m/imx8mm/platform.mk
@@ -50,7 +50,7 @@ BL31_SOURCES          +=      plat/imx/common/imx8_helpers.S                  \
                                ${IMX_GIC_SOURCES}
 
 USE_COHERENT_MEM       :=      1
-RESET_TO_BL31              :=      1
+RESET_TO_BL31                :=      0
 A53_DISABLE_NON_TEMPORAL_HINT := 0
 
 ERRATA_A53_835769      :=      1

```

2) Alternatively (**preferred**) we could not clear the arguments passed by SPL while keeping RESET_TO_BL31. For instance just define a new macro for imx8m (KEEP_FIRST_BOOTLOADER_ARGUMENTS) and continue to enter via reset.

```
diff --git a/bl31/aarch64/bl31_entrypoint.S b/bl31/aarch64/bl31_entrypoint.S
index ed058648f..f62b60520 100644
--- a/bl31/aarch64/bl31_entrypoint.S
+++ b/bl31/aarch64/bl31_entrypoint.S
@@ -72,11 +72,14 @@ func bl31_entrypoint
         * there's no argument to relay from a previous bootloader. Zero the
         * arguments passed to the platform layer to reflect that.
         * ---------------------------------------------------------------------
-    */
+ */
+#ifndef KEEP_FIRST_BOOTLOADER_ARGUMENTS
        mov     x20, 0
        mov     x21, 0
        mov     x22, 0
        mov     x23, 0
+#endif
+

```

### CODE REVIEW

**Comment by ldts** [comment]
@jforissier checkpatch seems to have a problem with _CFG_USE_DTB_OVERLAY (I didnt add this option btw)

**Comment by jforissier** [comment]
> @jforissier checkpatch seems to have a problem with _CFG_USE_DTB_OVERLAY (I didnt add this option btw)

Yes, checkpatch assumes there is only one possible prefix for configuration variables (the default being `CONFIG_` as in the Linux kernel, we [set it](https://github.com/OP-TEE/optee_os/blob/3.16.0/.checkpatch.conf#L13) to `CFG_` for OP-TEE). Let's ignore the warning.

File: lib/libutils/isoc/include/string.h [hunk 1]
```diff
@@ -30,6 +30,7 @@ char *strstr(const char *big, const char *little);
 char *strcpy(char *dest, const char *src);
 char *strncpy(char *dest, const char *src, size_t n);
 char *strrchr(const char *s, int i);
+char* strtok(char *str, const char *delims);
```
Comment (jforissier) [inline lib/libutils/isoc/include/string.h:33]: I'd rather have `strtok_r()` instead, so it can be easily used in thread context.
Style: `char *strtok(...);`.

File: lib/libutils/isoc/strtok.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 Foundries Ltd <jorge@foundries.io>
+ */
+#include <stdlib.h>
+#include <string.h>
+
+char* strtok(char *str, const char *delims)
```
Comment (jenswi-linaro) [inline lib/libutils/isoc/strtok.c:8]: Why not copy this function (or `strtok_r()` as suggested by @jforissier) from newlib instead?

File: lib/libutils/isoc/newlib/strtok_r.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#undef __STRICT_ANSI__
+#include "_ansi.h"
+#include <string.h>
+
+static char *
+_DEFUN (__strtok_r, (s, delim, lasts, skip_leading_delim),
```
Comment (jforissier) [inline lib/libutils/isoc/newlib/strtok_r.c:35]: The latest upstream version doesn't have this ugly `_DEFUN` anymore, I'd rather we use that instead (although we do have some older code too). https://github.com/bminor/newlib/blob/master/newlib/libc/string/strtok_r.c

Comment (ldts) [reply lib/libutils/isoc/newlib/strtok_r.c:35]: sure, let me pull and update.

**Comment by jforissier** [comment]
"libutils: Import strtok_r from newlib" is 
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

(if you don't mind and while you're at it, add parentheses after the function name in the subject and commit description: `strtok_r()` not `strtok_r`)

File: core/arch/arm/kernel/boot.c [hunk 4]
```diff
@@ -652,6 +655,10 @@ static int add_dt_overlay_fragment(struct dt_descriptor *dt, int ioffs)
 	int offs;
 	int ret;
 
+#ifdef CFG_OVERLAY_ADDR
+	if (!dt->is_overlay)
+		return ioffs;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/boot.c:669]: should return an `FDT_ERR_xxx` value, or -1 as done at line 671.
(note -1 is not really nice: it relates to `FDT_ERR_NOTFOUND`. At least it makes the implementation consistent.

File: core/arch/arm/kernel/boot.c [hunk 5]
```diff
@@ -1287,6 +1301,11 @@ static void init_primary(unsigned long pageable_part, unsigned long nsec_entry)
  */
 void __weak boot_init_primary_late(unsigned long fdt)
 {
+#ifdef CFG_OVERLAY_ADDR
+	struct dt_descriptor *dt = &external_dt;
+
+	dt->is_overlay = 0;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/boot.c:1454]: `external_dt.is_overlay = true;`

File: mk/config.mk [hunk 6]
```diff
@@ -439,22 +439,52 @@ CFG_DT ?= n
...
+# Use Case 2: CFG_EXTERNAL_DTB_OVERLAY, applies an overlay into an existing DTB
+#             overlay. If no valid overlay is found, create it.
+#
+#             The  DTB location (physical address) is provided either by
+#             1. boot argument arg2.
+#             2. CFG_DT_ADDR.
+#
+# Use Case 3: CFG_OVERLAY_ADDR, creates an overlay at the physical address
+#             provided with this config AND updates the DTB.
+#             i,e: SPL passes the U-boot DTB address in params; and the kernel
```
Comment (etienne-lms) [inline mk/config.mk:454]: prefer `CFG_OVERLAY_ADDR` is renamed to something like `CFG_NSEC_DTB_OVERLAY_ADDR`, that is with `_DTB_OVERLAY_ADDR` rather than only `_OVERLAY_ADDR`.

File: mk/config.mk [hunk 7]
```diff
@@ -439,22 +439,52 @@ CFG_DT ?= n
...
+#             CFG_DT_ADDR
+#
+#
 # A subsequent boot stage can then merge the generated overlay DTB into a main
 # DTB using the standard fdt_overlay_apply() method.
+
 CFG_EXTERNAL_DTB_OVERLAY ?= n
 CFG_GENERATE_DTB_OVERLAY ?= n
 
+ifeq ($(CFG_OVERLAY_ADDR),)
```
Comment (etienne-lms) [inline mk/config.mk:468]: `ifneq ($(CFG_OVERLAY_ADDR),y)`

File: core/arch/arm/kernel/boot.c [hunk 8]
```diff
@@ -1310,8 +1329,16 @@ void __weak boot_init_primary_late(unsigned long fdt)
...
 	} else {
+		/* use runtime init calls to update DTB */
 		init_tee_runtime();
 	}
+
+#ifdef CFG_OVERLAY_ADDR
+	release_external_dt();
+	dt->is_overlay = 1;
+	init_external_dt(CFG_OVERLAY_ADDR);
+	update_external_dt();
```
Comment (etienne-lms) [inline core/arch/arm/kernel/boot.c:1487]: complete the sequence with `release_external_dt()`.

File: mk/config.mk [hunk 9]
```diff
@@ -439,22 +439,52 @@ CFG_DT ?= n
...
+#             The  DTB location (physical address) is provided either by
+#             1. boot argument arg2.
+#             2. CFG_DT_ADDR.
+#
+# Use Case 3: CFG_OVERLAY_ADDR, creates an overlay at the physical address
+#             provided with this config AND updates the DTB.
+#             i,e: SPL passes the U-boot DTB address in params; and the kernel
+#                  overlay address is received in the config.
+#
+#             Condition: do NOT configure
```
Comment (etienne-lms) [inline mk/config.mk:457]: maybe more explicit as: `Exclusive with CFG_EXTERNAL_DTB_OVERLAY, CFG_GENERATE_DTB_OVERLAY and CFG_DT_ADDR`.

File: mk/config.mk [hunk 10]
```diff
@@ -439,22 +439,52 @@ CFG_DT ?= n
...
+
+ifneq ($(strip $(CFG_OVERLAY_ADDR)),)
+ifeq ($(CFG_EXTERNAL_DTB_OVERLAY),y)
+$(error Cannot implement OVERLAY_ADDR and EXTERNAL_DTB_OVERLAY)
+endif
+ifeq ($(CFG_GENERATE_DTB_OVERLAY),y)
+$(error Cannot implement OVERLAY_ADDR and GENERATE_DTB_OVERLAY)
+endif
+ifneq ($(strip $(CFG_DT_ADDR)),)
+$(error Cannot implement OVERLAY_ADDR and CFG_DT_ADDR)
```
Comment (etienne-lms) [inline mk/config.mk:484]: I think here `CFG_OVERLAY_ADDR` is still compatible with `CFG_DT_ADDR`. The later can point to the DTB loaded by earlier boot stage (as SPL) while `CFG_OVERLAY_ADDR` points to the DTB overlay binary generated by OP-TEE. Am I right?

File: core/arch/arm/kernel/boot.c [hunk 11]
```diff
@@ -997,13 +997,17 @@ static int get_nsec_memory_helper(void *fdt, struct core_mmu_phys_mem *mem)
...
-	addr_size = fdt_address_cells(fdt, 0);
-	if (addr_size < 0)
-		return 0;
-
-	len_size = fdt_size_cells(fdt, 0);
-	if (len_size < 0)
-		return 0;
+	if (IS_ENABLED(_CFG_USE_DTB_OVERLAY)) {
+		len_size = sizeof(paddr_t) / sizeof(uint32_t);
+		addr_size = sizeof(paddr_t) / sizeof(uint32_t);
```
Comment (etienne-lms) [inline core/arch/arm/kernel/boot.c:1011]: When `_CFG_USE_DTB_OVERLAY` is enabled and the target memory already stores a DTB, OP-TEE is expected to update this DTB. In that case, size/address cell counts should still be read from the DTB image.

**Review by etienne-lms — COMMENTED** [review]
Comments for commits "core: overlay: process DTB and generate DTB overlay" 
and "core: calculate size/address cells with overlay".

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"libutils: Import strtok_r() from newlib"

File: core/arch/arm/kernel/boot.c (hunk 8 above)
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1487]: this is taken care off by the boot final stuff - I think it is good as is.,

File: mk/config.mk (hunk 10 above)
Comment (ldts) [reply mk/config.mk:484]: um not at the moment (see entry_a64.S) copying the value to x20... I could make it compatible with a little bit more of work but I'd rather have that as a separate enhancement.

**Comment by clementfaure** [comment]
Hi @ldts,
Could you explain this change in ATF?
```
-RESET_TO_BL31              :=      1
+RESET_TO_BL31                :=      0
```

**Comment by ldts** [comment]
> Hi @ldts, Could you explain this change in ATF?
> 
> ```
> -RESET_TO_BL31              :=      1
> +RESET_TO_BL31                :=      0
> ```

Sorry yes, this deserves an explanation.

I haven't posted the change yet for review since I am looking a bit more into the full set of changes needed in SPL; having said that the imx8m TF-A changes discussion should be independent from this PR IMO: I just used imx8m since is the board I had at hand. 

on imx8m SPL boots TF-A so if the bootrom starts SPL at the right exception level I see no reason why SPL cant take the necessary steps to warm boot TF-A.

At the moment, the change that gets me to a working Linux shell (ie all cores initialized) just requires the following instead of disabling RESET_TO_BL31 (which will require a bit more investigation)...I dont know for sure but starting with RESET_TO_BL31 might have been a shortcut to avoid SPL changes (I might be wrong but this has been done in the past)

```
diff --git a/bl31/aarch64/bl31_entrypoint.S b/bl31/aarch64/bl31_entrypoint.S
index ed058648f..f62b60520 100644
--- a/bl31/aarch64/bl31_entrypoint.S
+++ b/bl31/aarch64/bl31_entrypoint.S
@@ -72,11 +72,14 @@ func bl31_entrypoint
         * there's no argument to relay from a previous bootloader. Zero the
         * arguments passed to the platform layer to reflect that.
         * ---------------------------------------------------------------------
-    */
+ */
+#ifndef KEEP_FIRST_BOOTLOADER_ARGUMENTS
        mov     x20, 0
        mov     x21, 0
        mov     x22, 0
        mov     x23, 0
+#endif
+

```

File: core/arch/arm/kernel/boot.c [hunk 12]
```diff
@@ -1153,6 +1293,9 @@ static void update_external_dt(void)
 	if (config_psci(dt))
 		panic("Failed to config PSCI");
 
+	if (reserve_platform_drivers_dt_nodes(dt))
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1296]: Right that is one way of doing it. I think a better way would be have macros in some i`mxX-reserve.h` files.

We could also add runtime support without having to change much (add a line to boot_init_primary_late)..we would just need to make sure the reserve_drivers array is fully populated after the driver has initialized.I can add support for that if you see value?

File: core/arch/arm/plat-imx/main.c [hunk 13]
```diff
@@ -115,6 +115,21 @@ void console_init(void)
 #endif
 }
 
+static const char *const reserved[] = {
+#if defined(CFG_NXP_CAAM)
+#if defined(CFG_MX8MM) || defined(CFG_MX8MQ) || defined(CFG_MX8MP)
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/plat-imx/main.c:120]: I think they could be extended to all iMX8m products..

File: core/arch/arm/kernel/boot.c (hunk 8 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/boot.c:1487]: That's right.

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "arm: boot: reserve devices for exclusive access".

Note: init functions playing with DT  in _boot.c_ are really growing. I think they would deserve their own source file. What do you think? Could be addressed aside or after this P-R.

File: core/arch/arm/kernel/boot.c [hunk 14]
```diff
@@ -1140,6 +1149,137 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+		if (dt->is_overlay)
+			offs = add_descriptor_node(dt, drivers[i]);
+		else
+#endif
+			offs = fdt_path_offset(dt->blob, drivers[i]);
+
+		if (offs < 0)
+			return -1;
+
+		if (!offs)
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1270]: you are right. this is a bug. if it was the overlay it would have been ok (it meant it had already been reserved) But not for a DTB. will remove this check. thanks!

File: core/arch/arm/kernel/boot.c [hunk 15]
```diff
@@ -1140,6 +1149,137 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+		/* OP-TEE handles first a DTB populated with a __symbols__ node
+		 * passed in parameters. Then, using those symbols, it generates
+		 * the overlay.
+		 */
+		goto handle;
+	}
+
+	offs = fdt_path_offset(dt->blob, "/__symbols__");
+	if (offs < 0)
+		goto handle;
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1230]: 100% right. will fix.

File: core/arch/arm/kernel/boot.c [hunk 16]
```diff
@@ -1140,6 +1149,137 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+
+	return offs;
+}
+#endif
+
+/* Guarantee that these drivers will not be used by either EL1 or EL0: set the
+ * device tree node status in the FDT and overlay to disabled and the
+ * secure-status to okay.
+ */
+static int reserve_platform_drivers_dt_nodes(struct dt_descriptor *dt)
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1205]: I did follow the pattern used on similar functions invoked from the same context (ie `update_external_dt()`). I think we should keep it as int.

File: mk/config.mk (hunk 7 above)
Comment (ldts) [reply mk/config.mk:468]: but CFG_OVERLAY_ADDR is an actual physical address

File: core/arch/arm/kernel/boot.c [hunk 17]
```diff
@@ -1140,6 +1149,137 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+		goto handle;
+	}
+
+	offs = fdt_path_offset(dt->blob, "/__symbols__");
+	if (offs < 0)
+		goto handle;
+
+	drvs = main_get_optee_reserved_drivers(&num_drivers);
+	if (!drvs || !num_drivers)
+		return 0;
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/arch/arm/kernel/boot.c:1234]: thanks, this would have introduced a nasty global regression.

File: core/arch/arm/kernel/boot.c (hunk 4 above)
Comment (ldts) [reply core/arch/arm/kernel/boot.c:669]: @etienne-lms um no I dont think so. If this is not the overlay when CFG_OVERLAY_ADDR is enabled, this could behave as if it was a DTB (ie, as if _CFG_USE_DTB_OVERLAY is not defined); and that code return offs.

File: core/arch/arm/kernel/boot.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/kernel/boot.c:1205]: Ok, but I think this is an inheritance from the first function that returned _libdft.h_ negative codes on error. Ok, let's consider moving to `TEE_Result` later, for all applicable functions.

**Comment by ldts** [comment]
@etienne-lms thanks a lot for all your comments. I did push a first update but will continue with your review (you had comments that I didnt look into yet)

**Comment by ldts** [comment]
even though all TF-A imx8m support is in Codeaurora I am proposing this change in trusted-firmware-a:
https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/14596

**Comment by ldts** [comment]
@clementfaure @etienne-lms any further comments?

**Comment by ldts** [comment]
@jforissier I realize everyone is busy with the release but just checking that this PR is just following its course and not forgotten? no rush from my end of course

**Comment by jforissier** [comment]
@ldts I doubt the review can be completed before the freeze for 3.17.0-rc1 (Friday), so this will probably have to go in later.

**Comment by ldts** [comment]
jfyi seems the TF-A PR is moving ahead.

**Comment by ldts** [comment]
@jforissier so what should we do with this PR?

**Comment by jforissier** [comment]
For commit "ci: azure: specify use case for CFG_OVERLAY_ADDR":
In the commit subject, s/CFG_OVERLAY_ADDR/CFG_NSEC_DTB_OVERLAY_ADDR/ then:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

As for the other unreviewed commits, I'd like opinions form @etienne-lms and @clementfaure who commented already.

**Comment by clementfaure** [comment]
Hello guys,
I've asked some folks in the BSP team to check what the impact of the ATF changes would be on imx8 bootflow (NXP internal tree). Depending on the result, we will use this feature or not.
@ldts If you want to go forward with this PR, you can remove `plat-imx: caam: iMX8m reserve job ring` as I cannot approve it for now.

File: core/arch/arm/plat-imx/main.c [hunk 18]
```diff
@@ -115,6 +115,21 @@ void console_init(void)
...
+	"sec_jr0",
```
Comment (clementfaure) [inline core/arch/arm/plat-imx/main.c:121]: The job ring to disable is not always `jr0`. The job ring to disable is `CFG_JR_INDEX` defined in `crypto_conf.mk`

**Comment by ldts** [comment]
@clementfaure ok thanks. I have removed the support for iMX from this PR. 
Since this is moving ahead https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/14596, adding support for the imx tf-a to pass the dtb with the symbols to op-tee should be trivial (as explained in this PR)...but no rush, I understand it might require some planning if agreed.

@jforissier I did add your Acked-by to the azure commit.

**Comment by ldts** [comment]
change has been merged

**Comment by ldts** [comment]
what is happening with this? when can a decision be made?
the alternative is to hardcode paths

**Comment by clementfaure** [comment]
Hi @ldts,
I don't have any feedback from BSP maintainers about the option you introduced in your TF-A pull-request.
Just one question : could the two methods to modify the DTB overlay coexist in optee-os ?

**Comment by ldts** [comment]
> Hi @ldts, I don't have any feedback from BSP maintainers about the option you introduced in your TF-A pull-request. Just one question : could the two methods to modify the DTB overlay coexist in optee-os ?

Right, the TF-A PR was merged some weeks ago so the functionality is upstream. But sure, both options could coexist: ie hardcoding paths or using the DTB. Just that hardcoding paths has a higher potential of failing (ie, an update could lead to a security issue if a path changes and OP-TEE cant reserve the device)

**Comment by ldts** [comment]
@jenswi-linaro @jbech-linaro would appreciate an answer on this.

@ricardosalveti we might have to carry this patch off-tree to keep our system secure. maybe we should tag it.

**Comment by jenswi-linaro** [comment]
@ldts, answer on what?

**Comment by ldts** [comment]
> @ldts, answer on what?

TL;DR 
1) Should OP-TEE hardcode paths to dt nodes or 2) should OP-TEE use __symbols"
if 1) I'll drop this PR
if 2) when can this PR be merged - what is missing.

**Comment by ldts** [comment]
> From what I've understood `2` is the preferred option. Any downside to that?

Platforms wanting to use 2 obviously must be able to pass the DTB to OP-TEE (ie, that requires a TF-A change on top of the one I already upstreamed for imx8. It is nothing major but someone needs to do it). If platforms cant pass a DTB then they wont be able to reserve - maybe for those platforms hardcoding the path could be an alternative (but this could be abused)

**Comment by clementfaure** [comment]
> Platforms wanting to use 2 obviously must be able to pass the DTB to OP-TEE (ie, that requires a TF-A change on top of the one I already upstreamed for imx8.

When you say that, you mean that you tested your TF-A patch on imx8m, right? You didn't literally add the support to imx8m in upstream, right?

> It is nothing major but someone in NXP needs to do it.

It's not guaranteed that NXP will add the support in the TF-A. I guess the BSP team will review this option during the next rebase.

Modifying the overlay with __symbols is a great thing to have, for sure. However, my main concern is that the feature requires to have the right platform support in TF-A ... If we enforce the use of `__symbols` to edit the DTB overlay, in the end, `CFG_EXTERNAL_DTB_OVERLAY=y` would require `RESET_TO_BL31_WITH_PARAMS=0` in the TF-A ?

**Comment by ldts** [comment]
> > Platforms wanting to use 2 obviously must be able to pass the DTB to OP-TEE (ie, that requires a TF-A change on top of the one I already upstreamed for imx8.
> 
> When you say that, you mean that you tested your TF-A patch on imx8m, right? You didn't literally add the support to imx8m in upstream, right?

I added generic support upstream to enable TF-A to pass the parameters to OP-TEE (applicable to the imx8 case)
https://github.com/ARM-software/arm-trusted-firmware/commit/25844ff728e4a0e5430ba2032457aba7b780a701

The TF-A changes required could be as simple as skipping the current hardcoding of the addresses when there is something being sent from TF-A (see below) I would have been open to help proposing alternatives at the time but I cant really spend more time on this (2 months is enough...as far as I am concerned I dont really care what they do now, no longer my problem, they should have engaged back then). 

```
diff --git a/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c b/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
index 049192880..66af6de5f 100644
--- a/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
+++ b/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c
@@ -176,7 +176,10 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
        /* Make sure memory is clean */
        mmio_write_32(BL32_FDT_OVERLAY_ADDR, 0);
        bl33_image_ep_info.args.arg3 = BL32_FDT_OVERLAY_ADDR;
-   bl32_image_ep_info.args.arg3 = BL32_FDT_OVERLAY_ADDR;
+ /* bl32 could be receiving a DTB on input while also asked to generate
+  * an overlay
+  */
+ bl32_image_ep_info.args.arg3 = arg1 > 0 ? arg1 : BL32_FDT_OVERLAY_ADDR;
 #endif
 #endif

```
which is what I used to test on imx8mm.

> 
> > It is nothing major but someone in NXP needs to do it.
> 
> It's not guaranteed that NXP will add the support in the TF-A. I guess the BSP team will review this option during the next rebase.
> 

sure. the BSP team might wish to continue hardcoding stuff. IMO is not very flexible and in this case it kind of limits what OP-TEE has to offer but that is the platform's choice.

> Modifying the overlay with __symbols is a great thing to have, for sure. However, my main concern is that the feature requires to have the right platform support in TF-A ... If we enforce the use of `__symbols` to edit the DTB overlay, in the end, `CFG_EXTERNAL_DTB_OVERLAY=y` would require `RESET_TO_BL31_WITH_PARAMS=0` in the TF-A ?

Kind of depend on each platform. But for iMX yes, it needs that so the param registers are not cleared when RESET_TO_BL31 is enabled.

**Comment by jenswi-linaro** [comment]
It seems that we need both options. If I understand this correctly, you're implementing option 2 here? If we take this then we can save option number 1 to whoever needs it enough to implement and upstream it.

**Comment by ldts** [comment]
> It seems that we need both options. If I understand this correctly, you're implementing option 2 here? If we take this then we can save option number 1 to whoever needs it enough to implement and upstream it.

right.

File: core/arch/arm/kernel/boot.c [hunk 19]
```diff
@@ -115,6 +115,15 @@ __weak unsigned long plat_get_aslr_seed(void)
 	return 0;
 }
 
+/* May be overridden in plat-$(PLATFORM)/main.c */
+__weak const char * const *main_get_optee_reserved_drivers(size_t *num)
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/boot.c:119]: I'd prefer `plat_get_reserved_drivers()`.

File: core/arch/arm/kernel/boot.c [hunk 20]
```diff
@@ -1133,6 +1142,138 @@ static int mark_tzdram_as_reserved(struct dt_descriptor *dt)
...
+static int reserve_platform_drivers_dt_nodes(struct dt_descriptor *dt)
+{
+	static const char **drivers;
+	static size_t num_drivers;
+	const char * const *drvs = { NULL };
+	const char *path = NULL;
+	unsigned int i = 0;
+	int offs = 0;
+	int len = 0;
+#ifdef CFG_NSEC_DTB_OVERLAY_ADDR
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/boot.c:1214]: Please use `if (IS_ENALBED(CFG_NSEC_DTB_OVERLAY_ADDR)` instead of ifdefs where possible.

**Review by jenswi-linaro — COMMENTED** [review]
Comments for the commit "arm: boot: reserve devices for exclusive access"

**Comment by ldts** [comment]
@jenswi-linaro before investing more time on this, is this going to be merged or are users going to end up hardcoding paths? I'd rather not waste the reviewers time (or my time) otherwise.

**Comment by jforissier** [comment]
> I believe that at least @jforissier and I agree that this is useful.

Yes. And I have merged the `strtok_r()` commit.
